src/librustc_session/config.rs

```diff
@@ -2435,43 +2459,103 @@ fn parse_borrowck_mode(dopts: &DebuggingOptions, error_format: ErrorOutputType)
         let mut parts = arg.splitn(2, '=');
-        let name = parts.next().unwrap_or_else(||
-            early_error(error_format, "--extern value must not be empty"));
-        let location = parts.next().map(|s| s.to_string());
+        let name = parts
+            .next()
+            .unwrap_or_else(|| early_error(error_format, "--extern value must not be empty"));
+        let path = parts.next().map(|s| s.to_string());
+
+        let mut name_parts = name.splitn(2, ':');
+        let first_part = name_parts.next();
+        let second_part = name_parts.next();
+        let (options, name) = match (first_part, second_part) {
+            (Some(opts), Some(name)) => (Some(opts), name),
+            (Some(name), None) => (None, name),
+            (None, None) => early_error(error_format, "--extern name must not be empty"),
+            _ => unreachable!(),
+        };
 
-        let entry = externs
-            .entry(name.to_owned())
-            .or_default();
+        let entry = externs.entry(name.to_owned());
 
+        use std::collections::btree_map::Entry;
 
-        entry.locations.insert(location.clone());
+        let entry = if let Some(path) = path {
+            // --extern prelude_name=some_file.rlib
+            match entry {
+                Entry::Vacant(vacant) => {
+                    let mut files = BTreeSet::new();
+                    files.insert(path);
+                    vacant.insert(ExternEntry::new(ExternLocation::Files(files)))
+                }
+                Entry::Occupied(occupied) => {
+                    let ext_ent = occupied.into_mut();
+                    match ext_ent {
+                        ExternEntry { location: ExternLocation::Files(files), .. } => {
+                            files.insert(path);
+                        }
+                        _ => early_error(
+                            error_format,
+                            "--extern with `=path` cannot be specified with \
+                             --extern without `=path`",
+                        ),
+                    }
+                    ext_ent
+                }
+            }
+        } else {
+            // --extern prelude_name
+            match entry {
+                Entry::Vacant(vacant) => {
+                    vacant.insert(ExternEntry::new(ExternLocation::ExternPrelude))
+                }
+                Entry::Occupied(_) => {
+                    early_error(error_format, "--extern without `=path` can only be specified once")
+                }
+            }
+        };
+
+        let mut is_private_dep = false;
+        let mut add_prelude = true;
+        if let Some(opts) = options {
+            if !is_unstable_enabled {
+                early_error(
+                    error_format,
+                    "the `-Z unstable-options` flag must also be passed to \
+                     enable `--extern options",
+                );
+            }
+            for opt in opts.split(',') {
```

Comment:
I find it visually a little more explicit if there is a different separator. Otherwise, the last "option" becomes special — the name which can't appear anywhere else in the option list.  It probably doesn't matter to the handful of people who ever need to write this, but I like it to be more explicit.

I added a test, though it seemed a bit awkward.